### PR TITLE
Neural Machine Translation with Megatron Transformer Models (Tensor Parallel and Tarred Datasets Only)

### PR DESCRIPTION
**Pull request by MaximumEntropy**
# What does this PR do ?

Adds support to train Neural Machine Translation models with Megatron Transformer models that support tensor parallelism.

**Collection**: NLP

# Changelog 
- Add a `MegatronNMT` class that inherits from `MegatronLMEncoderDecoderModel` and overrides appropriate functions and calls classmethods from `MTEncDecModel`
- Change appropriate methods in `MTEncDecModel` to classmethods so they can be called in `MegatronNMTModel`.
- Add a config file and training script.

# Usage

```python
examples/nlp/machine_translation/megatron_nmt_training.py -cn aayn_base_megatron model.tensor_model_parallel_size=2
```

# Before your PR is "Ready for review"
**Pre checks**:
- [ ] Make sure you read and followed [Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md)
- [ ] Did you write any new necessary tests?
- [ ] Did you add or update any necessary documentation?
- [ ] Does the PR affect components that are optional to install? (Ex: Numba, Pynini, Apex etc)
  - [ ] Reviewer: Does the PR have correct import guards for all optional libraries?
  
**PR Type**:
- [ ] New Feature
- [ ] Bugfix
- [ ] Documentation

If you haven't finished some of the above items you can still open "Draft" PR.

## Who can review?

Anyone in the community is free to review the PR once the checks have passed. 
[Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md) contains specific people who can review PRs to various areas.

# Additional Information
* Related to # (issue)

### CODE REVIEW

File: examples/nlp/machine_translation/megatron_nmt_training.py [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+from nemo.utils import logging
+from nemo.utils.exp_manager import StatelessTimer, exp_manager
+
+
+@hydra_runner(config_path="conf", config_name="aayn_base_megatron")
+def main(cfg) -> None:
+    logging.info("\n\n************** Experiment configuration ***********")
+    logging.info(f'\n{OmegaConf.to_yaml(cfg)}')
+
+    megatron_amp_o2 = cfg.model.get('megatron_amp_O2', False)
```
Comment (michalivne) [inline examples/nlp/machine_translation/megatron_nmt_training.py:37]: We should move all the shared initialization part to a utility in the utility refactor. Adding here as a comment.

File: nemo/collections/nlp/data/machine_translation/machine_translation_dataset.py [hunk 2]
```diff
@@ -414,7 +414,7 @@ def __init__(
         self._dataset = wd.WebDataset(urls=text_tar_filepaths, nodesplitter=None)
 
         if shuffle_n > 0:
-            self._dataset = self._dataset.shuffle(shuffle_n)
+            self._dataset = self._dataset.shuffle(shuffle_n, initial=shuffle_n)
```
[the remark this answers is not in the document]
Comment (michalivne) [reply nemo/collections/nlp/data/machine_translation/machine_translation_dataset.py:417]: @ericharper right in what sense? Unlike np.shuffle which is in-place, this function returns a WebDataset object, so should be ok.

File: nemo/collections/nlp/models/language_modeling/megatron_lm_encoder_decoder_model.py [hunk 3]
```diff
@@ -407,8 +407,9 @@ def predict_step(self, batch: Any, batch_idx: int, dataloader_idx: Optional[int]
         logging.info(f"response: {response}")
         return response
 
-    def decode(self, tokens_enc, enc_mask, num_tokens_to_generate, enc_input=None):
+    def decode(self, tokens_enc, enc_mask, num_tokens_to_generate, enc_input=None, tokenizer=None):
         # TODO: move method into a class inside MegatronTokenLevelEncoderDecoderModule (?)
```
Comment (michalivne) [inline nemo/collections/nlp/models/language_modeling/megatron_lm_encoder_decoder_model.py:411]: Remove comment

File: nemo/collections/nlp/models/language_modeling/megatron_lm_encoder_decoder_model.py [hunk 4]
```diff
@@ -407,8 +407,9 @@ def predict_step(self, batch: Any, batch_idx: int, dataloader_idx: Optional[int]
         logging.info(f"response: {response}")
         return response
 
-    def decode(self, tokens_enc, enc_mask, num_tokens_to_generate, enc_input=None):
+    def decode(self, tokens_enc, enc_mask, num_tokens_to_generate, enc_input=None, tokenizer=None):
```
Comment (michalivne) [inline nemo/collections/nlp/models/language_modeling/megatron_lm_encoder_decoder_model.py:410]: Why is tokenizer can be given as an argument?

File: nemo/collections/nlp/models/machine_translation/megatron_nmt_model.py [hunk 5]
```diff
@@ -0,0 +1,416 @@
...
+
+__all__ = ["MegatronNMTModel"]
+
+
+class MegatronNMTModel(MegatronLMEncoderDecoderModel):
+    """
+    Megatron NMT training
+    """
+
+    def __init__(self, cfg: DictConfig, trainer: Trainer):
```
Comment (michalivne) [inline nemo/collections/nlp/models/machine_translation/megatron_nmt_model.py:48]: Should we use the existing p-tuning here instead of a custom soft-prompt?

File: nemo/collections/nlp/models/machine_translation/megatron_nmt_model.py [hunk 6]
```diff
@@ -0,0 +1,416 @@
...
+            make_vocab_size_divisible_by=self._cfg.get('make_vocab_size_divisible_by', 128),
+            tensor_model_parallel_size=self._cfg.get('tensor_model_parallel_size', 1),
+        )
+
+    def training_step(self, batch, batch_idx):
+        # Need to squeze dim 0 for tarred datasets since things are pre-batched and we ask the dataloader for batch size 1.
+        batch = [x.squeeze(dim=0) if x.ndim == 3 else x for x in batch]
+        hashitem = batch[0].sum().cpu().numpy().tolist()
+        assert hashitem not in self.train_hash_buffer
+        self.train_hash_buffer.add(hashitem)
```
Comment (michalivne) [inline nemo/collections/nlp/models/machine_translation/megatron_nmt_model.py:154]: Why do we need to hash the training sets in order to verify uniqueness? Is this required or was left as part of debugging?

File: nemo/collections/nlp/models/machine_translation/megatron_nmt_model.py [hunk 7]
```diff
@@ -0,0 +1,416 @@
...
+
+        # This returns the averaged loss across data-parallel groups.
+        reduced_loss = super().validation_step(batch, batch_idx)
+        tokens_enc, tokens_dec, loss_mask, labels, enc_mask, dec_mask = self.process_batch(batch)
+        predicted_tokens_ids, _ = self.decode(
+            tokens_enc,
+            enc_mask,
+            tokens_enc.size(1)
+            + self._cfg.max_generation_delta,  # Generate up to src-length + max generation delta. TODO: Implement better stopping when everything hits <EOS>.
+            tokenizer=self.decoder_tokenizer,
```
Comment (michalivne) [inline nemo/collections/nlp/models/machine_translation/megatron_nmt_model.py:169]: ok, this is why tokenizer is passed.

**Review by michalivne — COMMENTED** [review]
Great contribution! See comments

File: nemo/collections/nlp/models/language_modeling/megatron_lm_encoder_decoder_model.py (hunk 3 above)
Comment (MaximumEntropy) [reply nemo/collections/nlp/models/language_modeling/megatron_lm_encoder_decoder_model.py:411]: Done.

File: examples/nlp/machine_translation/megatron_nmt_training.py (hunk 1 above)
Comment (MaximumEntropy) [reply examples/nlp/machine_translation/megatron_nmt_training.py:37]: Agreed.

File: nemo/collections/nlp/models/language_modeling/megatron_lm_encoder_decoder_model.py (hunk 4 above)
Comment (MaximumEntropy) [reply nemo/collections/nlp/models/language_modeling/megatron_lm_encoder_decoder_model.py:410]: This is because in T5 style models, we could get away with always accessing `self.tokenizer` to do the IDs -> text conversion, but for NMT, we now have `self.encoder_tokenizer` and `self.decoder_tokenizer`.

File: nemo/collections/nlp/models/machine_translation/megatron_nmt_model.py (hunk 6 above)
Comment (MaximumEntropy) [reply nemo/collections/nlp/models/machine_translation/megatron_nmt_model.py:154]: Sorry, this needs to be removed. It was part of debugging. Done.

File: nemo/collections/nlp/data/machine_translation/machine_translation_dataset.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (MaximumEntropy) [reply nemo/collections/nlp/data/machine_translation/machine_translation_dataset.py:417]: Apparently webdatasets does some sort of lazy shuffling to speedup how fast you get your initial batches and setting `initial` to be a higher value forces things to be closer to a global shuffle at the expense of your first few batches getting slower. I thought I would test it to see if it helps debug our weird multilingual learning curves.

File: nemo/collections/nlp/models/machine_translation/megatron_nmt_model.py [hunk 8]
```diff
@@ -0,0 +1,416 @@
...
+        if self._train_dl is not None and self._validation_dl is not None:
+            return
+        self.build_train_valid_test_datasets()
+        self.setup_training_data(self._cfg.train_ds)
+        self.setup_validation_data(self._cfg.validation_ds)
+        self.setup_test_data(self._cfg.test_ds)
+
+    def _build_tokenizer(self):
+        # Instantiates tokenizers and register to be saved with NeMo Model archive
+        # After this call, ther will be self.encoder_tokenizer and self.decoder_tokenizer
```
[the remark this answers is not in the document]
Comment (MaximumEntropy) [reply nemo/collections/nlp/models/machine_translation/megatron_nmt_model.py:95]: Fixed.

File: nemo/collections/nlp/models/machine_translation/megatron_nmt_model.py (hunk 5 above)
Comment (MaximumEntropy) [reply nemo/collections/nlp/models/machine_translation/megatron_nmt_model.py:48]: Not sure I understand. There is no p-tuning or soft-prompt related code here no?

**Review by ericharper — APPROVED** [review]
LGTM. Thanks!

File: nemo/collections/nlp/models/machine_translation/megatron_nmt_model.py (hunk 5 above)
Comment (michalivne) [reply nemo/collections/nlp/models/machine_translation/megatron_nmt_model.py:48]: I meant for multi-lingual support. But could be added in a separated PR.

File: examples/nlp/machine_translation/noisy_channel_reranking.py [hunk 9]
```diff
@@ -296,7 +296,12 @@ def main():
             src_lines = []
             for line in src_f:
                 src_lines.append(line.strip().split('\t'))
-            assert len(all_reverse_scores) == len(all_lm_scores) == len(all_forward_scores) == len(src_lines)
+            assert len(all_reverse_scores) == len(all_lm_scores) == len(all_forward_scores) == len(src_lines), (
```
Comment (michalivne) [inline examples/nlp/machine_translation/noisy_channel_reranking.py:299]: Remove printing out or add meaningful message

**Review by michalivne — COMMENTED** [review]
LGTM! see minor comments

**Review by michalivne — APPROVED** [review]
LGTM!
